xml/encoder_test.go

```diff
@@ -0,0 +1,447 @@
+package xml_test
+
+import (
+	"bytes"
+	"testing"
+
+	"github.com/awslabs/smithy-go/xml"
+)
+
+func TestEncoder(t *testing.T) {
+	encoder := xml.NewEncoder()
+
+	obj := encoder.RootElement()
+	obj.Key("stringKey").String("stringValue")
+	obj.Key("integerKey").Integer(1024)
+	obj.Key("floatKey").Float(3.14)
+
+	vk, closeFn := obj.Key("foo").NestedElement()
+	vk.Key("byteSlice").String("Zm9vIGJhcg==")
+
+	if closeFn != nil {
+		closeFn()
+	}
+
+	e := []byte(`<stringKey>stringValue</stringKey><integerKey>1024</integerKey><floatKey>3.14</floatKey><foo><byteSlice>Zm9vIGJhcg==</byteSlice></foo>`)
+	if a := encoder.Bytes(); bytes.Compare(e, a) != 0 {
+		t.Errorf("expected %+q, but got %+q", e, a)
+	}
+
+	if a := encoder.String(); string(e) != a {
+		t.Errorf("expected %s, but got %s", e, a)
+	}
+}
+
+func TestEncodeAttribute(t *testing.T) {
+	encoder := xml.NewEncoder()
+	obj := encoder.RootElement()
+
+	obj.Key("payload", func(t *xml.TagMetadata) {
+		t.AttributeName = "attrkey"
+		t.AttributeValue = "value"
+	}).Null()
```

Comment:
Instead of creating a new `TagMetadata` each iterating to apply the functional options, you could create teh value once, then allow it to be filled. The `AttributeName`/`AttributeValue` could be changed to a map/string.

This would be more consistent with the "common functional argument pattern.

```go
type TagMetadata struct {
       NamespaceURI string
       NamespacePrefix String

       Attributes map[string]string
}

obj.Key("payload", func(t *xml.TagMetadata) {
     t.Attributes["attrKey"] = "value"
})
```